Date: 02/09/2018

Version: 1.0

Environment: Python 3.6 and Jupyter notebook

Libraries used: 
* re (for regular expression, included in Anaconda Python 3.6) 
* json (to convert given data into json format, included in Anaconda Python 3.6) 


## 1.  Import libraries and Load Data

In [52]:
# Imports the libraries used in Assignment task1
import re    
import json

#Loading the file  
with open('29416000.dat','r') as infile:
    text = infile.read()

# 2. Data Analysing or Visualization

This section is done in the background to determine which is the most occuring attribute and also to determine different versions/patterns of the same attribute. 

<B>import pandas as pd</B>

<B>attributes_df=pd.DataFrame(data=re.findall('[\n]\s*(_?[A-Za-z]+\s*?[_]?[A-Za-z]+?)[:]',text)) </B>
#The regular expression fetches all the Attributes in a file which are starting from Alphabets(uppercase or lowecase),Underscore,blank space and followed by ':' previous char should be a newline.

<B>count=attributes_df[0].value_counts().rename('attritube_freq')</B>
#This gives individual character frequency count.

<B>attributes_df = attributes_df.merge(count.to_frame(),left_on=0,right_index=True)</B>
#Merging it to form a new dataframe

<B>unq=attributes_df[0].unique()</B>
#unique attributes in a file which helped to determine the different versions of the attribute for instance, JOB_T and TITLE for description.

<B>new_df=attributes_df[attributes_df.attritube_freq>=65]</B>
#Neglecting all the attributes which have count less than 65(fixed by looking at the dataframe frequency values)  because those are the words picked up by the regular expression which are not exactly attributes(beacause the file contains the same pattern which are not attributes.

##  3.Program Flow (Json file creation)

#### 3.1 Regular Expression Compilation using re.complie
The block below compiles regular expressions pattern into regular expression object. It reduces the compilation time.

Syntax:re.compile(pattern, flags=0)

* Variations in Location attribute: LOCS,LOCATONS,LOC,JOB_LOC,_LOC,LOCATION <br/><br/>
* Variations in Title attribute:_TTL,TITLES,JOB_T,title <br/><br/>
* Variations in Description attribute: _description,DESCRIPTION,JOB_DESCRIPTION,JOB_DESC,JOb DESCRIPTION <br/><br/>
* Variations in About company attribute attribute: _info,ABOUT COMPANY,ABOUT,about_comany,COMAPNYS_INFO <br/><br/>
* Variations in Start Date attribute: DATES,DATE_START,START_DA,START DATE,start_date <br/><br/>
* Variations in Qualifications attribute: qualifications,QUALIFS,QUALIFICATION,EQ_QUALS,REQUIRED QUALIFICATIONS,QUALIFICATIONS <br/><br/>
* Variations in Application deadline attribute: APPLICATION_DEADL,APPLICATIONS_DL,DEADLINE,DEAD_LINE,deadline,DEADLINES <br/><br/>
* Variations in Resposibility attribute: RESP,resposibilitiy,RESPOSIBILTIY, JOB RESPOSIBILTIES ,JOB_RESPS <br/><br/>
* Variations in Procedure attribute: PROCEDURE,procedure,JOB_PROCS,JOB_PROC,PROCDEDURES <br/><br/>
* Variations in Salary attribute: JOB_SAL,SALARY,salary,Salary <br/><br/>
* Variations in Remuneration attribute: Remuneration, Remuneration.<br/><br/>


All these Variations are complied by writing a regular expression for each which is shown in the next block.

In [53]:
Loc=re.compile('_?LOCS?A?T?I?O?N?S?:|JOB_LOC:') #Regex fetches all the locations, Similary for all the attributes
Title=re.compile('_TTL:|TITLES?:|JOB_T|title:')
Job=re.compile('J?O?B?\s?_?DESCR?I?P?T?I?O?N?:|_description:|job_desc:',re.DOTALL)
About=re.compile('ABOUT?\s?C?O?M?P?A?N?Y?S?_?I?N?F?O?:|about_company|_info:|COMPANYS_INFO:',re.DOTALL)
Date=re.compile('DATES?_?S?T?A?R?T?:|START\s?_?D?A?T?E?:|start_date:')
Qualify=re.compile('QUALIF?S?I?C?A?T?I?O?N?S?:|qualifications:|REQUIRED QUALIFICATIONS:|REQ_QUALS:')
Deadline=re.compile('APPLICATION_DE?A?D?L?:|DEAD_?LINES?:|deadline:')
Resp=re.compile('J?O?B?_?\s?RESPO?N?S?I?B?I?L?I?T?Y?I?E?S?:|responsibilities:')
Proc=re.compile('PROCEDURES?:|JOB_PROCS?:',re.IGNORECASE)
Salary=re.compile('J?O?B?_?SALA?R?Y?:|\nSalary:',re.IGNORECASE)
Remuneration=re.compile('REMUNERATION:',re.IGNORECASE)
remove=re.compile('REMUNERATION/|START DATE/|OPEN TO/')#used to remove key which has '/'at the end (as it contains no value)

##### Checking the count of each  Attributes in all the job postings before modification

re.findall - Returns all the non overlapping matches of the passed string as a list. 

Syntax: re.findall(pattern, string)¶

In [54]:
# Finding all the patterns
location_C=re.findall(Loc,text)
Title_c=re.findall(Title,text)
Job_c=re.findall(Job,text)
About_c=re.findall(About,text)
Date_c=re.findall(Date,text)
Qualify_c=re.findall(Qualify,text)
Deadline_c=re.findall(Deadline,text)
Resp_c=re.findall(Resp,text)
Proc_c=re.findall(Proc,text)
Salary_c=re.findall(Salary,text)
Remuneration_c=re.findall(Remuneration,text)
Id_c=re.findall('ID:',text)

In [55]:
#printing the coount of each attribute before modification.
print('Total number of IDs:',len(Id_c))
print('Total number of Locations:',len(location_C))
print('Total number of Title:',len(Title_c))
print('Total number of Job Description:',len(Job_c))
print('Total number of About Company:',len(About_c))
print('Total number of Start Date:',len(Date_c))
print('Total number of Qualification:',len(Qualify_c))
print('Total number of Procedure:',len(Proc_c))
print('Total number of Remuneration:',len(Remuneration_c))
print('Total number of Salary:',len(Salary_c))
print('Total number of Resposibility:',len(Resp_c))
print('Total number of Deadline:',len(Deadline_c))


Total number of IDs: 30503
Total number of Locations: 28329
Total number of Title: 28376
Total number of Job Description: 28684
Total number of About Company: 28252
Total number of Start Date: 28410
Total number of Qualification: 28495
Total number of Procedure: 28377
Total number of Remuneration: 10025
Total number of Salary: 18462
Total number of Resposibility: 28411
Total number of Deadline: 28334


#### 3.2. Replacing the file with the common key attribute 
Replaces the string which matches the pattern compiled in the 3.1 section.The block below replaces every pattern matched with a common string.

Syntax:re.sub(pattern, repl, string, count=0, flags=0)

<font size=3 color="blue">Why replacing with commom attribute ? </font><br/>

I am doing this because all the different versions of the attribute have to be normalised before passing it to json or xml. Instead of replacing it in the later stages, I felt replacing in the beginning would be easier to fetch values of each attribute. 



In [56]:
# Replacing with the attributes with common names
modified_file=re.sub(Loc,'LOCATION:',text)
modified_file=re.sub(Title,'TITLE:',modified_file)
modified_file=re.sub(About,r'ABOUT:',modified_file)
modified_file=re.sub(Job,r'DESCRIPTON:',modified_file)
modified_file=re.sub(Date,r'STARTDATE:',modified_file)
modified_file=re.sub(Qualify,r'QUALIFICATION:',modified_file)
modified_file=re.sub(Deadline,r'DEADLINE:',modified_file)
modified_file=re.sub(Resp,r'RESPONSIBILITY:',modified_file)
modified_file=re.sub(Proc,r'PROCEDURE:',modified_file)
modified_file=re.sub(Salary,r'SALARY:',modified_file)
modified_file=re.sub(Remuneration,r'REMUNERATION:',modified_file)
modified_file=re.sub(remove,r'',modified_file)# removing the attributes which deosn't have value.
modified_file=re.sub('REMUNERATION:','SALARY:',modified_file)

##### Fetching the counts after replacing the keywords

Fetching the counts for each attribute after modification and ensuring the counts before and after the replacing matches.

In [57]:
location_C=re.findall('LOCATION:',modified_file)
Title_c=re.findall('TITLE:',modified_file)
Job_c=re.findall('DESCRIPTON:',modified_file)
About_c=re.findall('ABOUT:',modified_file)
Date_c=re.findall('STARTDATE:',modified_file)
Qualify_c=re.findall('QUALIFICATION:',modified_file)
Deadline_c=re.findall('DEADLINE:',modified_file)
Resp_c=re.findall('RESPONSIBILITY:',modified_file)
Proc_c=re.findall('PROCEDURE:',modified_file)
Salary_c=re.findall('SALARY:',modified_file)
Id_c=re.findall('ID:',modified_file)

In [58]:
print('Total number of IDs:',len(Id_c))
print('Total number of Locations:',len(location_C))
print('Total number of Title:',len(Title_c))
print('Total number of Job Description:',len(Job_c))
print('Total number of About Company:',len(About_c))
print('Total number of Start Date:',len(Date_c))
print('Total number of Qualification:',len(Qualify_c))
print('Total number of Procedure:',len(Proc_c))
print('Total number of Salary:',len(Salary_c))
print('Total number of Resposibility:',len(Resp_c))
print('Total number of Deadline:',len(Deadline_c))

Total number of IDs: 30503
Total number of Locations: 28329
Total number of Title: 28376
Total number of Job Description: 28684
Total number of About Company: 28252
Total number of Start Date: 28410
Total number of Qualification: 28495
Total number of Procedure: 28377
Total number of Salary: 28487
Total number of Resposibility: 28411
Total number of Deadline: 28334


#### 3.3 Splitting the file based on the identified delimitter<br/>


<font size=3 color="blue">Why we need to split ?</font>

Splitting would be necessary as we need to capture information for each job posting.  

The delimitter is identified as `'------------------------------'` and deleting the last index as at the end there is also delimitter which eventually gives one empty string at the end.

In [59]:
split_modified_file=modified_file.split('------------------------------')# splits and stores it as a list os job postings
del split_modified_file[len(split_modified_file)-1] # deleting last index 

#### 3.4 Extracting All the Attributes

re.findall method extracts the pattern given as a input and return a list. If more than one group is captured it returns as list of tuples.  

Syntax: re.findall(pattern, string, flags=0)

In [60]:
list_dict=[] # stores list of dictionaries 
for each in split_modified_file: # running a loop for each Job postings
    job_dic={}   # initializing the empty dictionary
    keys=re.findall(r'([A-Z]+):',each) # extract all keys matching the pattern
    for i in range(len(keys)):#running a loop to extract values between two keys               
        if i<len(keys)-1: # condtion is put to avoid out of range, as 'i+1' goes out of index when it reached last attribute
            job_dic[keys[i]]=re.search(keys[i]+':'+'(.*?)'+keys[i+1],each,re.DOTALL).group(1)# fetches key,value pair dynamically.
        else:
            job_dic[keys[i]]=re.search(keys[i]+':'+'(.*)',each,re.DOTALL).group(1)# fetches last attribute key,value pair
    list_dict.append(job_dic) # finally appends to list of dictionaries

#### 3.5 Handling Inner divisions of attributes

As observed RESPONSIBILITY,DESCRIPTION,QUALIFICATION have subparts and hence needs division accordingly.

In [61]:
for each in list_dict:
    for i,j in each.items():
        if len(re.findall(r'(\n-.*?\s*;)',j)) !=0: # checks if any attribute has subdivion by checking if there are bullet points'-' and ends with ';' 
            if i=='RESPONSIBILITY':
                resp=j.replace('-','').strip().split(';')# spiltting by ';' ,replacing '-' by null and stripping blank spaces.
                obj_resp={'Responsibility':resp} # Assigning it to dictionary
                each[i]=obj_resp # assigning back to attribute.
            elif i=='DESCRIPTION':# The above steps are repeated for description and qualification
                desc=j.replace('-','').strip().split(';')
                obj_desc={'Description':desc}
                each[i]=obj_desc
            elif i=='QUALIFICATION':
                qualf=j.replace('-','').strip().split(';')
                obj_qualf={'Description':qualf}
                each[i]=obj_qualf

#### 3.6 Creating Dictionary objects 

This helps in creating Dictionaries inside Dictionary which is easier to dump into json method to covert into json file.

In [62]:
obj_listing={'Listing':list_dict} # Adding one dictionary to other
obj_listings={'Listings':obj_listing}

#### 3.7 Writing the output to json file

Its uses json.dump method which converts dictionary of dictionary to json file.

In [63]:
with open('29416000.json', 'w') as out:
    json.dump(obj_listings, out) 

## 4. Program Flow (Xml file creation)

The program takes input as dictionary of dictionaries which is created in the previous section and loops over each dictionary and appends angular brackets{<''>... </>} for each key,value pair.

I did not wanted to hard code values like 'required qualications' etc and hence using the same key for the headings inside   

In [64]:
result=[] # empty list to store each output, Looping over each dictionaries and attaching angular brackets to Attributes and Values
for i,j in obj_listings.items():    
    result.append('<'+i+'>')
    for x,y in j.items():
        for _ in y:
            result.append('<'+x+'>')
            for k,v  in _.items():              
                temp=type(v)
                if temp!=dict:                   
                   result.append('<'+k+'>'+v+'</'+k+'>') 
                else:
                    for m,n in v.items():
                        result.append('<'+m+'>')
                        for each in n:                            
                            result.append('<'+m+'>'+each+'</'+m+'>')
                        result.append('</'+m+'>')
            result.append('</'+x+'>')       
    result.append('</'+i+'>')

#### 4.1 Joining the result list into a string. 
File contains '\n' and hence replacing it with null.

In [65]:
finalxml=''.join(result) # converts list into string
replace=finalxml.replace('\n','') # when converted it adds '\n' and hence removing it by using replace


#### 4.2 Writing the output to XML file

Its uses file.write method which converts string into .xml file

In [66]:
File = open('29416000.xml','w') 
File.write(str(replace))

84138348

We can say that ID is identifier and it is present in all the Job postings.Hence Normalizing all attributes with respect to ID which we can infer the below following.


In [67]:
#Printing the percentages of Each attribute
print('Percetage of IDs:',len(Id_c)/len(Id_c)*100)
print('Percetage of Locations:',len(location_C)/len(Id_c)*100)
print('Percetage of Title:',len(Title_c)/len(Id_c)*100)
print('Percetage of Job Description:',len(Job_c)/len(Id_c)*100)
print('Percetage of About Company:',len(About_c)/len(Id_c)*100)
print('Percetage of Start Date:',len(Date_c)/len(Id_c)*100)
print('Percetage of Qualification:',len(Qualify_c)/len(Id_c)*100)
print('Percetage of Procedure:',len(Proc_c)/len(Id_c)*100)
print('Percetage of Salary:',len(Salary_c)/len(Id_c)*100)
print('Percetage of Resposibility:',len(Resp_c)/len(Id_c)*100)
print('Percetage of Deadline:',len(Deadline_c)/len(Id_c)*100)

Percetage of IDs: 100.0
Percetage of Locations: 92.87283218044126
Percetage of Title: 93.02691538537194
Percetage of Job Description: 94.03665213257713
Percetage of About Company: 92.62039799363997
Percetage of Start Date: 93.13837983149197
Percetage of Qualification: 93.41704094679211
Percetage of Procedure: 93.03019375143428
Percetage of Salary: 93.39081401829328
Percetage of Resposibility: 93.14165819755434
Percetage of Deadline: 92.88922401075304


## 5. Summary

The provided Job Postings Dataset contains <b> 30503 </b>different job postings. I have identified totally <b>11</b> key attributes for each job posting. As per the statistics shown there may be missing key attribitues in each job posting.The precentage of each attribute present in whole text file provided is as shown below.

* Percetage of IDs: 100.0
* Percetage of Locations: 92.87283218044126
* Percetage of Title: 93.02691538537194
* Percetage of Job Description: 94.03665213257713
* Percetage of About Company: 92.62039799363997
* Percetage of Start Date: 93.13837983149197
* Percetage of Qualification: 93.41704094679211
* Percetage of Procedure: 93.03019375143428
* Percetage of Salary: 93.39081401829328
* Percetage of Resposibility: 93.14165819755434
* Percetage of Deadline: 92.88922401075304

The steps performed in Data Extraction and Manipulation are as follows:

1. Writing a regex to extract Attributes
2. Analysing the attributes by its frequency and Uniqueness to get different versions of attributes.
3. Writing a regex for each attribute and replacing it with single word.
4. Extracting the text between two Atrributes.
5. Creating json objects to finally dump as json file.
6. Constructing XML strings by adding tags with the help of already created json object.
7. Writing to file. 

## 5. References 

* https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html
 
* https://docs.python.org/3/library/re.html

* https://docs.python.org/3/library/json.html
